src/ol-ext/format/featurehash.js

```diff
@@ -0,0 +1,940 @@
+/**
+ * @fileoverview This file provides an OpenLayers format for encoding
+ * and decoding features for use in permalinks.
+ *
+ * The code is based on Stéphane Brunner's URLCompressed format:
+ * <https://github.com/sbrunner/OpenLayers-URLCompressed>
+ *
+ * TODO
+ * - The OpenLayers-URLCompressed format has options where the user
+ *   can define attribute and style transformers. This is currently
+ *   not supported by the ngeo format.
+ */
+goog.provide('ngeo.format.FeatureHash');
+
+goog.require('goog.asserts');
+goog.require('ol.Feature');
+goog.require('ol.format.TextFeature');
+goog.require('ol.geom.GeometryLayout');
+goog.require('ol.geom.GeometryType');
+goog.require('ol.geom.LineString');
+goog.require('ol.geom.MultiLineString');
+goog.require('ol.geom.MultiPoint');
+goog.require('ol.geom.MultiPolygon');
+goog.require('ol.geom.Point');
+goog.require('ol.geom.Polygon');
+goog.require('ol.style.Circle');
+goog.require('ol.style.Fill');
+goog.require('ol.style.Stroke');
+goog.require('ol.style.Style');
+goog.require('ol.style.Text');
+
+
+/**
+ * @enum {string}
+ */
+ngeo.format.FeatureHashStyleType = {
+  LINE_STRING: 'LineString',
+  POINT: 'Point',
+  POLYGON: 'Polygon'
+};
+
+
+/**
+ * @type {Object.<ol.geom.GeometryType, ngeo.format.FeatureHashStyleType>}
+ * @private
+ */
+ngeo.format.FeatureHashStyleTypes_ = {};
+
+ngeo.format.FeatureHashStyleTypes_[ol.geom.GeometryType.LINE_STRING] =
+    ngeo.format.FeatureHashStyleType.LINE_STRING;
+ngeo.format.FeatureHashStyleTypes_[ol.geom.GeometryType.POINT] =
+    ngeo.format.FeatureHashStyleType.POINT;
+ngeo.format.FeatureHashStyleTypes_[ol.geom.GeometryType.POLYGON] =
+    ngeo.format.FeatureHashStyleType.POLYGON;
+ngeo.format.FeatureHashStyleTypes_[ol.geom.GeometryType.MULTI_LINE_STRING] =
+    ngeo.format.FeatureHashStyleType.LINE_STRING;
+ngeo.format.FeatureHashStyleTypes_[ol.geom.GeometryType.MULTI_POINT] =
+    ngeo.format.FeatureHashStyleType.POINT;
+ngeo.format.FeatureHashStyleTypes_[ol.geom.GeometryType.MULTI_POLYGON] =
+    ngeo.format.FeatureHashStyleType.POLYGON;
+
+
+
+/**
+ * @constructor
+ * @param {ngeox.format.FeatureHashOptions=} opt_options Options.
+ * @extends {ol.format.TextFeature}
+ */
+ngeo.format.FeatureHash = function(opt_options) {
+  goog.base(this);
+
+  var options = goog.isDef(opt_options) ? opt_options : {};
+
+  /**
+   * @type {number}
+   * @private
+   */
+  this.accuracy_ = goog.isDef(options.accuracy) ?
+      options.accuracy : ngeo.format.FeatureHash.ACCURACY_;
+
+  /**
+   * @type {number}
+   * @private
+   */
+  this.prevX_ = 0;
+
+  /**
+   * @type {number}
+   * @private
+   */
+  this.prevY_ = 0;
+
+};
+goog.inherits(ngeo.format.FeatureHash, ol.format.TextFeature);
+
+
+/**
+ * Characters used to encode the data.
+ * ~'() will be used as separator.
```

Comment:
Separators to separate coordinates and properties and key/value pairs within properties blocks.